### PR TITLE
add @ManualAuthorization annotation for non-standard endpoints

### PR DESCRIPTION
**Pull request by apucher**
Several endpoints in pinot aren't served well by the AuthFilter's heuristics for extracting table names. In particular, this includes `/tableConfigs` and `/schema` endpoints that accept opaque payloads without a table or schema name in the parameters. This causes problems with fine-grained access control, such as non-admin users creating new tables via self-serve.

This PR adds a new annotation `@ManualAuthorization` for REST endpoints, which allows developers to skip the default authorization and deserialize payloads before manually invoking authorization, e.g. via `AccessControlUtils.validatePermissions()`. This annotation comes with obvious risks and should be used sparingly, as it enables requests to bypass most of the AuthFilter.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentUploadDownloadRestletResource.java [hunk 1]
```diff
@@ -139,24 +138,12 @@ public class PinotSegmentUploadDownloadRestletResource {
...
-      AccessControl accessControl = _accessControlFactory.create();
-      hasDataAccess = accessControl.hasDataAccess(httpHeaders, tableName);
-    } catch (Exception e) {
-      throw new ControllerApplicationException(LOGGER,
-          "Caught exception while validating access to table: " + tableName, Response.Status.INTERNAL_SERVER_ERROR, e);
-    }
-    if (!hasDataAccess) {
-      throw new ControllerApplicationException(LOGGER, "No data access to table: " + tableName,
-          Response.Status.FORBIDDEN);
-    }
```
Comment (sajjad-moradi) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentUploadDownloadRestletResource.java:147]: We shouldn't delete the manual authentication here. Although the method is GET, but we need to only allow ppl who are authorized to be able to download the data.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java [hunk 2]
```diff
@@ -193,7 +198,7 @@ private String getQueryResponse(String query, @Nullable SqlNode sqlNode, String
 
     // Validate data access
     AccessControl accessControl = _accessControlFactory.create();
-    if (!accessControl.hasDataAccess(httpHeaders, rawTableName)) {
+    if (!accessControl.hasAccess(rawTableName, AccessType.READ, httpHeaders, endpointUrl)) {
```
Comment (sajjad-moradi) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:201]: Let's not change this. Although we're reading data and AccessType.READ looks appropriate, we need to only allow ppl who are authorized to be able to query the data that might contain sensitive info like user information. It's different than reading table config or some other cluster configs that might be ok with weaker authentication level like AccessType.READ.

**Review by sajjad-moradi — CHANGES_REQUESTED** [review]

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentUploadDownloadRestletResource.java (hunk 1 above)
Comment (apucher) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentUploadDownloadRestletResource.java:147]: This is the same strictness as before. The endpoint provides a table name, therefore, we'll perform auth based on headers and table name. The only change is removing the call to a deprecated method.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/access/AccessControl.java [hunk 3]
```diff
@@ -51,7 +52,8 @@ public interface AccessControl {
    * @param endpointUrl the request url for which this access control is called
    * @return whether the client has permission
    */
-  default boolean hasAccess(String tableName, AccessType accessType, HttpHeaders httpHeaders, String endpointUrl) {
+  default boolean hasAccess(@Nullable String tableName, AccessType accessType, HttpHeaders httpHeaders,
```
[the remark this answers is not in the document]
Comment (apucher) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/access/AccessControl.java:55]: We already scenarios where `null` is passed in here "by accident" if `tableName` (schemaName) happens to come out to null. This merely acknowledges that reality. Do you know of an existing implementation that has issues with this?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java (hunk 2 above)
Comment (apucher) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java:201]: Similar to the above, this has similar strictness to the original implementation. The difference here is removing the use of a deprecated method.

On a tangent, this resource shouldn't be in the controller in the first place, rather, it should be controlled by the broker's access control.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/access/AccessControl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (walterddr) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/access/AccessControl.java:57]: +1. table name can already be null. also the access control doesn't distinguish between table or schema name

**Review by sajjad-moradi — APPROVED** [review]

**Pull request merged**